OrbitCaptureClient/CaptureEventProcessorTest.cpp

```diff
@@ -0,0 +1,614 @@
+  CaptureEventProcessor event_processor(&listener);
+
+  CaptureEvent event;
+  CallstackSample* callstack_sample = event.mutable_callstack_sample();
+  callstack_sample->set_pid(1);
+  callstack_sample->set_tid(3);
+  Callstack* callstack = callstack_sample->mutable_callstack();
+  callstack->add_pcs(14);
+  callstack->add_pcs(15);
+  callstack_sample->set_timestamp_ns(100);
+
+  CallStack actual_call_stack;
+  EXPECT_CALL(listener, OnUniqueCallStack).Times(1).WillOnce(SaveArg<0>(&actual_call_stack));
+  CallstackEvent actual_call_stack_event;
+  EXPECT_CALL(listener, OnCallstackEvent).Times(1).WillOnce(SaveArg<0>(&actual_call_stack_event));
+
+  event_processor.ProcessEvent(event);
+
+  EXPECT_EQ(actual_call_stack_event.time(), callstack_sample->timestamp_ns());
+  EXPECT_EQ(actual_call_stack_event.thread_id(), callstack_sample->tid());
+  EXPECT_EQ(actual_call_stack_event.callstack_hash(), actual_call_stack.GetHash());
+  ASSERT_EQ(actual_call_stack.GetFramesCount(), callstack->pcs_size());
+  for (size_t i = 0; i < actual_call_stack.GetFramesCount(); ++i) {
+    EXPECT_EQ(actual_call_stack.GetFrame(i), callstack->pcs(i));
+  }
+}
+
+TEST(CaptureEventProcessor, WillOnlyHandleUniqueCallstacksOnce) {
+  MockCaptureListener listener;
+  CaptureEventProcessor event_processor(&listener);
+  std::vector<CaptureEvent> events;
+
+  CaptureEvent event_1;
+  CallstackSample* callstack_sample_1 = event_1.mutable_callstack_sample();
+  callstack_sample_1->set_pid(1);
+  callstack_sample_1->set_tid(3);
+  Callstack* callstack_1 = callstack_sample_1->mutable_callstack();
+  callstack_1->add_pcs(14);
+  callstack_1->add_pcs(15);
+  callstack_sample_1->set_timestamp_ns(100);
+
+  CaptureEvent event_2;
+  CallstackSample* callstack_sample_2 = event_2.mutable_callstack_sample();
+  callstack_sample_2->set_pid(2);
+  callstack_sample_2->set_tid(4);
+  Callstack* callstack_2 = callstack_sample_2->mutable_callstack();
+  callstack_2->add_pcs(14);
+  callstack_2->add_pcs(15);
+  callstack_sample_2->set_timestamp_ns(200);
+
+  CallStack actual_call_stack;
+  EXPECT_CALL(listener, OnUniqueCallStack).Times(1).WillOnce(SaveArg<0>(&actual_call_stack));
+  CallstackEvent actual_call_stack_event_1;
+  CallstackEvent actual_call_stack_event_2;
+  EXPECT_CALL(listener, OnCallstackEvent)
+      .Times(2)
+      .WillOnce(SaveArg<0>(&actual_call_stack_event_1))
+      .WillOnce(SaveArg<0>(&actual_call_stack_event_2));
+
+  event_processor.ProcessEvent(event_1);
+  event_processor.ProcessEvent(event_2);
+
+  EXPECT_EQ(actual_call_stack_event_1.time(), callstack_sample_1->timestamp_ns());
+  EXPECT_EQ(actual_call_stack_event_1.thread_id(), callstack_sample_1->tid());
+  EXPECT_EQ(actual_call_stack_event_1.callstack_hash(), actual_call_stack.GetHash());
+  ASSERT_EQ(actual_call_stack.GetFramesCount(), callstack_1->pcs_size());
+  for (size_t i = 0; i < actual_call_stack.GetFramesCount(); ++i) {
+    EXPECT_EQ(actual_call_stack.GetFrame(i), callstack_1->pcs(i));
+  }
```

Comment:
This equality-comparison-code appears twice in this test and also in other tests. What about having a function for that?

Reply:
Done. Thanks for the idea.